From 372c9e4a7c103e3737c34cb2bba6a913f609da58 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Mon, 17 Mar 2025 18:18:13 +0100
Subject: [PATCH] pwm: stm32-lp: don't rely on regmap to manage the clock

The MFD core (parent) driver uses devm_regmap_init_mmio_clk() to enable
or disable the clock upon each register access through regmap API.

It's not possible when using CONFIG_PREEMPT_RT, to manage the clock
from within regmap, when clocksource driver is used. The IRQ requires
to access the register, hence calls clk_[en|dis]able. It takes a
spin_lock which isn't allowed in hard irq context (IRQF_TIMER flag):
-> stm32_clkevent_lp_irq_handler
 -> regmap_write
  -> clk_enable
   -> rt_spin_lock
It's highlighted in logs by "BUG: scheduling while atomic..." stack dumps.

This is a precursor rework to replace devm_regmap_init_mmio_clk() use
in the MFD core parent. It requires to manage the clock directly here.

There's no functional change intended here. Instead, call pm_runtime
get/put method to manage the clock, before/after accessing the
registers.

Change-Id: I8576405f50d97a521cd052ecb87a3f9b65e5f6b4
Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445399
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Olivier MOYSAN <olivier.moysan@foss.st.com>
---
 drivers/pwm/pwm-stm32-lp.c | 35 ++++++++++++++++++++++-------------
 1 file changed, 22 insertions(+), 13 deletions(-)

--- a/drivers/pwm/pwm-stm32-lp.c
+++ b/drivers/pwm/pwm-stm32-lp.c
@@ -188,9 +188,16 @@ static int stm32_pwm_lp_apply(struct pwm
 	dty = prd * state->duty_cycle;
 	dty = DIV_ROUND_CLOSEST_ULL(dty, state->period);
 
+	if (!cstate.enabled) {
+		/* enable clock to drive PWM counter */
+		ret = pm_runtime_resume_and_get(chip->dev);
+		if (ret)
+			return ret;
+	}
+
 	ret = regmap_read(priv->regmap, STM32_LPTIM_CFGR, &cfgr);
 	if (ret)
-		return ret;
+		goto err;
 
 	/*
 	 * When there are several channels, they share the same prescaler and reload value.
@@ -199,17 +206,12 @@ static int stm32_pwm_lp_apply(struct pwm
 	if (!stm32_pwm_lp_update_allowed(priv, pwm->hwpwm)) {
 		ret = regmap_read(priv->regmap, STM32_LPTIM_ARR, &arr);
 		if (ret)
-			return ret;
-
-		if ((FIELD_GET(STM32_LPTIM_PRESC, cfgr) != presc) || (arr != prd - 1))
-			return -EBUSY;
-	}
+			goto err;
 
-	if (!cstate.enabled) {
-		/* enable clock to drive PWM counter */
-		ret = pm_runtime_resume_and_get(chip->dev);
-		if (ret)
-			return ret;
+		if ((FIELD_GET(STM32_LPTIM_PRESC, cfgr) != presc) || (arr != prd - 1)) {
+			ret = -EBUSY;
+			goto err;
+		}
 	}
 
 	if ((FIELD_GET(STM32_LPTIM_PRESC, cfgr) != presc) ||
@@ -304,6 +306,11 @@ static int stm32_pwm_lp_get_state(struct
 	u64 tmp;
 	int ret;
 
+	/* Enable the clock to access registers */
+	ret = pm_runtime_resume_and_get(chip->dev);
+	if (ret < 0)
+		return ret;
+
 	regmap_read(priv->regmap, STM32_LPTIM_CR, &val);
 	enabled = !!FIELD_GET(STM32_LPTIM_ENABLE, val);
 	if (priv->num_cc_chans) {
@@ -319,8 +326,10 @@ static int stm32_pwm_lp_get_state(struct
 	/* Keep PWM counter clock refcount in sync with PWM initial state */
 	if (state->enabled) {
 		ret = pm_runtime_resume_and_get(chip->dev);
-		if (ret < 0)
+		if (ret < 0) {
+			pm_runtime_put_sync_suspend(chip->dev);
 			return ret;
+		}
 	}
 
 	regmap_read(priv->regmap, STM32_LPTIM_CFGR, &val);
@@ -344,7 +353,7 @@ static int stm32_pwm_lp_get_state(struct
 	tmp = (tmp << presc) * NSEC_PER_SEC;
 	state->duty_cycle = DIV_ROUND_CLOSEST_ULL(tmp, rate);
 
-	return 0;
+	return pm_runtime_put_sync_suspend(chip->dev);
 }
 
 static const struct pwm_ops stm32_pwm_lp_ops = {
